### ISSUE TITLE
Add History Chart to all ERC20 tokens and Ether [Samoa]

### ISSUE THREAD

**Issue by colourfreak** [AlphaWallet/alpha-wallet-ios#2629]
Zeplin:
https://zpl.io/amPBKqv

<img width="941" alt="Screenshot 2021-03-10 at 15 06 15" src="https://user-images.githubusercontent.com/51817359/110641323-2b90be00-81b2-11eb-8904-56f21d838af5.png">

My suggestions:
- Coingecko API
- Tradingview API

**Comment by colourfreak** [comment]
In some cases, we can use the chart from the underlying token. For example, when you have cWBTC token, we can display the chart from WBTC. So this is all about linking Defi tokens with those underlying assets. 

Do you have any idea how we can play with this? @hboon

**Comment by colourfreak** [comment]
![2021-09-08 07 42 00](https://user-images.githubusercontent.com/51817359/132526119-acb55979-d065-4b67-8f6b-b077d4f5f34d.png)

**Comment by hboon** [comment]
> In some cases, we can use the chart from the underlying token. For example, when you have cWBTC token, we can display the chart from WBTC. So this is all about linking Defi tokens with those underlying assets.
> 
> Do you have any idea how we can play with this? @hboon

We'll have to manually map this out by hand and then hardcode that mapping into the app. Some of this might be cross-chains. So the mapping is a list of pairs like:

* cWBTC <contract> (Ethereum mainnet) = WBTC <contract> (Ethereum mainnet)
* native (Ethereum mainnet) = WETH <contract> (Optimism mainnet)
* ...

We'll need to specify if it's the native token or the contract (address) and the chain it is on for each mapping

We have to be pretty sure that the fiat value are really the same for each of this though.

Do you want to make a list first (specify the contract if you can find it)? Then we can build from there.

**Comment by colourfreak** [comment]
If we are planning to do this manually, let's think about how to kill two birds with one stone. There are many other issues we can add on while doing this manual process.

**For example:**
- We can specify if a token goes to group DEFI, Staking, Governance, etc.
- What is the network: later on, we could write a script that automatically places this small network icon I am placing manually at the moment. So all you will have to do is to upload the token symbol once, then it will detect a network and place an icon in the corner. I think the Rainbow app does the same. 
- etc... ideas?

**Comment by hboon** [comment]
> We can specify if a token goes to group DEFI, Staking, Governance, etc.

I would reply again maybe tomorrow, but sending you this link in case you want to research. I'd prefer if we can ride on token lists like https://tokenlists.org, otherwise it'd be a lot of non-coding building and maintenance work for the lists.

**Comment by hboon** [comment]
> If we are planning to do this manually, let's think about how to kill two birds with one stone. There are many other issues we can add on while doing this manual process.

Yes and no. Seems like they are discrete tasks, but I agree they can fall under some umbrella in the name of improving token taxonomy or something like that. They are discrete because the data for each one seems to not overlap. Let's think about these, but hopefully we can do it one by one and only the easy one first (aligning the UI and bug fixes are still top priority):

> We can specify if a token goes to group DEFI, Staking, Governance, etc.

Yes, categorization of tokens into groups like would be best implemented with something like https://tokenlists.org where we either can expend little effort to building it or we can get many people to build it with us.

> What is the network: later on, we could write a script that automatically places this small network icon I am placing manually at the moment. So all you will have to do is to upload the token symbol once, then it will detect a network and place an icon in the corner

We don't need a script here. We already know the chain the contract/token is on when we display it in the app, so the app can do that overlay. Would you create an issue for this if you want to proceed this way. There is a logistic issue because there is only 1 repo version available and we'll need to figure out a way to support app versions which does the overlay and earlier versions which doesn't.

> etc... ideas?

Token lists can also be used in the add custom token screen to help auto-populate fields when the user enters a contract or search by name

In a similar manner, but for chains, rather than tokens, https://chainlist.org can be helpful to present data when user add custom chains in the UI

Quite a couple of areas, but focus :)

If we want to continue to explore these separately from price (this issue), would you help create/move these to a separate issue? We should definitely implement them at some point. They are quite basic to the core of wallets, just that there are even more core things that ought to work better first.

**Comment by colourfreak** [comment]
1. Tokens with the same prices like ETH and CETH
2. Group it belongs to ie. Defi, Assets, Governance
3. Network ie. Polygon, Optimism
4. The same tokens but on different networks ie. LINK on Polygon and Mainnet
5. V1 and v2 of the same token ie. AAVE had version 1 and 2 for their tokens, if we can link them together we could possible requests users to move funds to newer version

This is what comes to my mind. Feel free to add more.

**Comment by hboon** [comment]
Try this format:

```
[
  {
    "address": "0xabcd",
    "chainId": 1,
    "group": "Assets",
    "decimals" : 18,
    "name" : "Zoo Token",
    "symbol" : "ZOOT",
    "isOldVersion": true,
    "alternative": {
      "address": "0xdef00",
      "chainId": 100
    }
  },
  {
    "address": "0xdef00",
    "chainId": 100,
    "group": "Assets",
    "decimals" : 18,
    "name" : "Zoo Token",
    "symbol" : "ZOOT",
    "isOldVersion": false,
  },

  ... more
]
```

`address`, `chainId`, `group`, `decimals`, `name` and `symbol` are mandatory. `alternative` is optional (like the second example in the list). They are used for looking to another token when they represent absolutely the same thing. `isOldVersion` is only needed when there are multiple version os a token, to indicate which is the newer version — flag all old versions with `isOldVersion = true` and the newest version with `isOldVersion = false`.

**Comment by colourfreak** [comment]
I requested Aleks, let's see.

**Comment by colourfreak** [comment]
@hboon Can you please add [@YouPlusWe](https://github.com/YouPlusWe) in this issue.

**Comment by colourfreak** [comment]
This is brief for Aleksandar.

**Summary:**
We need a database we can reuse while grouping tokens in AlphaWallet, link similar tokens together and so on.

If you install iOS AlphaWallet app, you will find that some tokens ex. cWBTC is missing chart and price info. We use Coingecko API to get this data. So logically if we are able to get data for WBTC token, we should be able to use the same data for DEFI tokens where WBTC is an underlying asset. 

There is an important comment provided by our Product Manager and iOS Lead Developer 
https://github.com/AlphaWallet/alpha-wallet-ios/issues/2629#issuecomment-916607668

This is the format we are looking for.
[
  {
    "address": "0xabcd",
    "chainId": 1,
    "group": "Assets",
    "decimals" : 18,
    "name" : "Zoo Token",
    "symbol" : "ZOOT",
    "isOldVersion": true,
    "alternative": {
      "address": "0xdef00",
      "chainId": 100
    }
  },
  {
    "address": "0xdef00",
    "chainId": 100,
    "group": "Assets",
    "decimals" : 18,
    "name" : "Zoo Token",
    "symbol" : "ZOOT",
    "isOldVersion": false,
  },

  ... more
https://github.com/AlphaWallet/alpha-wallet-ios/issues/2629#issuecomment-917837196

**Groups:**
- Assets (all ERC20 tokens + Ether that has a history chart and price)
- NFTs (All collectibles)
- DEFI (for all savings, liquidity pools, loans, debt obligations)
- Governance (for all DAO tokens, if you have this token, you can vote, participate in DAO etc.)
- Utilities (all other tokens)

Where to take this data from?
* https://chainlist.org/
* https://tokenlists.org/
* https://polygonscan.com/
* https://www.bscscan.com/
* https://hecoinfo.com/
* https://zerion.io/
* https://blockscout.com/
* https://optimistic.etherscan.io/
* https://explorer.avax.network/
* https://explorer.fantom.network/assets

**Comment by JamesSmartCell** [comment]
Hi @colourfreak, does this issue help #3172 ? Although if it's currently not showing and is on main net then coingekko may not be listing that info.

Additionally, we may be able to get offer/price listing for NFT from OpenSea. Should we include this information at some point?

This will be difficult as there is often a difference between listing and offer prices (eg Chicken Derby 15000eth vs 70eth).

**Comment by colourfreak** [comment]
Can you please assign to @hieronymus4826?

**Comment by hboon** [comment]
> @hboon Can you please add [@YouPlusWe](https://github.com/YouPlusWe) in this issue.
> Can you please assign to @hieronymus4826?

Do you mean to assign to both of them?

They'll need to respond here before we can assign to them (because GitHub).

It's not very clear  to me what they are supposed to do though. I might have missed it. But could you describe or point out (if it has been written above) what we/they are supposed to do?

**Comment by colourfreak** [comment]
They are trying to write the scraper.

**Comment by hieronymus4826** [comment]
After the meeting with @YouPlusWe on Friday, we think the best known thing for Defi would be: https://blog.unmarshal.io/unmarshal-lp-token-position-api-b55a598cbc65

It is constantly updated with new integrations and has an active network.

**Comment by hboon** [comment]
Created #3423 since we are mixing 2 issues into one. The price history display is already implemented.

**Comment by colourfreak** [comment]
> Created #3423 since we are mixing 2 issues into one. The price history display is already implemented.

Do we have an issue for the second part? I mean the one when I am trying to describe each token? https://github.com/AlphaWallet/alpha-wallet-ios/issues/2629#issuecomment-918291148

**Comment by hboon** [comment]
@colourfreak not specifically. If you need it not ti support #3423, then would you help create one an issue describe what it’s for?
